packages/plugins/apollo-usage-report/src/reporter.ts

```diff
@@ -0,0 +1,189 @@
+    maxBatchDelay: number;
+    maxBatchUncompressedSize: number;
+    maxTraceSize: number;
+    exportTimeout: number;
+  };
+
+  #reportsBySchema: Record<string, OurReport> = {};
+  #nextSendAfterDelay?: ReturnType<typeof setTimeout>;
+  #sending: Promise<unknown>[] = [];
+
+  constructor(
+    options: ApolloUsageReportOptions,
+    yoga: YogaServer<Record<string, unknown>, Record<string, unknown>>,
+    logger: YogaLogger,
+  ) {
+    this.#logger = logger;
+    this.#yoga = yoga;
+    this.#options = {
+      ...options,
+      maxBatchDelay: options.maxBatchDelay ?? 20 * 60 * 1000, // 20min
+      maxBatchUncompressedSize: options.maxBatchUncompressedSize ?? 4 * 1024 * 1024, // 4mb
+      maxTraceSize: options.maxTraceSize ?? 10 * 1024 * 1024, // 10mb
+      exportTimeout: options.exportTimeout ?? 30 * 1000, // 60s
+    };
+    this.#reportHeaders = {
+      graphRef: getGraphRef(options),
+      hostname: options.hostname ?? getEnvVar('HOSTNAME') ?? '',
+      uname: options.uname ?? '', // TODO: find a cross-platform way to get the uname
+      runtimeVersion: options.runtimeVersion ?? '',
+      agentVersion: options.agentVersion || `graphql-yoga@${yoga.version}`,
+    };
+  }
+
+  addTrace(schemaId: string, options: Parameters<OurReport['addTrace']>[0]) {
+    const report = this.#getReport(schemaId);
+    report.addTrace(options);
+
+    if (
+      this.#options.alwaysSend ||
+      report.sizeEstimator.bytes >= this.#options.maxBatchUncompressedSize!
+    ) {
+      return this.sendReport(schemaId);
+    }
+
+    this.#nextSendAfterDelay ||= setTimeout(() => this.flush(), this.#options.maxBatchDelay);
+
+    return;
+  }
+
+  async flush() {
+    return Promise.allSettled([
+      ...this.#sending, // When flushing, we want to also wait for previous traces to be sent, because it's mostly used for clean up
+      ...Object.keys(this.#reportsBySchema).map(schemaId => this.sendReport(schemaId)),
+    ]);
+  }
+
+  async sendReport(schemaId: string) {
+    const sending = this.#sendReport(schemaId);
+    this.#sending.push(sending);
+    sending.finally(() => this.#sending?.filter(p => p !== sending));
+    return sending;
+  }
+
+  async #sendReport(schemaId: string) {
+    const {
+      fetchAPI: { fetch, CompressionStream, ReadableStream },
+    } = this.#yoga;
+    const report = this.#reportsBySchema[schemaId];
+    if (!report) {
+      throw new Error(`No report to send for schema ${schemaId}`);
+    }
+
+    if (this.#nextSendAfterDelay != null) {
+      clearTimeout(this.#nextSendAfterDelay);
+    }
+
+    delete this.#reportsBySchema[schemaId];
+    report.endTime = dateToProtoTimestamp(new Date());
+    report.ensureCountsAreIntegers();
+
+    const validationError = Report.verify(report);
+    if (!validationError) {
+      throw new TypeError(`Invalid report: ${validationError}`);
+    }
+
+    const encodedReport = Report.encode(report).finish();
+    const compressedReport = new ReadableStream({
```

Comment:
The same readable stream might not be used multiple times. Worth to double-check

Reply:
Oh yeah, I must probably do this for each try